DOC Add extra information on how to customise an file HTTP response (#9514)
silverstripe/silverstripe-framework: docs/en/02_Developer_Guides/14_Files/03_File_Security.md

## Configuring protected assets

> In most cases, developers can just interact with File and Image objects without worrying about how

Reviewer: ```suggestion
In most cases, developers can interact with File and Image objects without worrying about how
```

> Silverstripe CMS resolve file names or responds to requests. Some advanced use cases may occasionally

Reviewer: ```suggestion
Silverstripe CMS resolves file names or responds to requests. Some advanced use cases may occasionally
```

> Silverstripe CMS resolve file names or responds to requests. Some advanced uses case will occasionally

Reviewer: ```suggestion
Silverstripe CMS resolve file names or responds to requests. Some advanced use cases may occasionally
```

Author: Fixed

> require developers to adjust the HTTP response to file request.

Reviewer: ```suggestion
require developers to adjust the HTTP response for file requests.
```

> Most of the routing logic for serving Files is controlled via the `AssetStore` interface. The main

Reviewer: ```suggestion
Most of the routing logic for serving Files is controlled via the `AssetStore` interface. The default
```

Author: Fixed

    Pragma: 'no-cache'
```

### Configuring file HTTP response code

When a user tries to access a file that exists, but for which they do not have access,
Silverstripe CMS will return a "404 Not found" response rather than a "403 Denied" to
avoid revealing the existence of the file.

> You can customise the response codes for various types of request via

Reviewer: ```suggestion
You can customise the response codes for various types of requests via
```

Author: Fixed

configuration flags on `FlysystemAssetStore`.

```yml

> # Add this code snippet to your YML configuration

Reviewer: ```suggestion
```

I don't think we need to say things like this, there are YAML snippets throughout our documentation - devs are pretty familiar with what it is by the point they're this deep in our docs.

Author: Fixed.

SilverStripe\Filesystem\Flysystem\FlysystemAssetStore:

>   denied_response_code: 403 # This default to 404

Reviewer 2: ```suggestion
  denied_response_code: 403 # The default for this is 404
```

Author: Fixed

  missing_response_code: 404
  redirect_response_code: 302
  permanent_redirect_response_code: 301
```

### Updating a file HTTP response before it's sent back to the browser

> `silverstripe/assets` 1.6 and above allows you to intercept the file HTTP response

Author: Need to make sure we update this line if the related PR doesn't get merged in time for the 4.6 release.

Reviewer: It's merged

before it's sent to the client by applying an `Extension` to `FlysystemAssetStore`.

To achieve this create an `Extension` and implement the `updateResponse` method.

```php
<?php

namespace App\MySite;

use SilverStripe\Control\HTTPResponse;
use SilverStripe\Core\Extension;

class AssetStoreExtension extends Extension
{
    /**
     * @param HTTPResponse $response Update this object to modify the response

>      * @param string $asset   Path of the request minus the `assets` prefix

Reviewer 2: alignment is a little out

Reviewer: I don't see what you mean - can you suggest an edit to show?

Reviewer 2: Sorry, meant the line above is a little out `@param HTTPResponse    $response`

Author: Fixed.

     * @param array $context  This array contains some resolution information from
     *                        FlysystemAssetStore. It may be empty. It may contain
     *                        a `visibility` key to say if we are serving a public
     *                        or protected file. It may contain a `parsedFileID`
     *                        detailing how FlysystemAssetStore has resolved $asset
     */

>     public function updateResponse($response, $asset, $context)

Reviewer: ```suggestion
    public function updateResponse(HTTPResponse $response, string $asset, array $context): void
```

I think we should start documenting with PHP 7 syntax where possible

Author: Fixed.

    {
        // Do something to the response
    }
}
```

> To enable the extension, add this snippet to you YML configuration file.

Reviewer: ```suggestion
Enable the extension with YML configuration:
```

Author: Fixed.
